From 45d28b481a32b5ca5b594790c77012f34fb878b6 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:08:46 -0400
Subject: [powerpc] ftrace: bugfix for test_24bit_addr

Message-id: <1410545655-205645-298-git-send-email-dzickus@redhat.com>
Patchwork-id: 94264
O-Subject: [RHEL7 PATCH 297/626] powerpc/ftrace: bugfix for test_24bit_addr
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit a95fc58549e8f462e560868a16b1fa97b12d5db6
Author: Liu Ping Fan <kernelfans@gmail.com>
Date:   Wed Feb 26 10:23:01 2014 +0800

    powerpc/ftrace: bugfix for test_24bit_addr

    The branch target should be the func addr, not the addr of func_descr_t.
    So using ppc_function_entry() to generate the right target addr.

    Signed-off-by: Liu Ping Fan <pingfank@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/ftrace.c b/arch/powerpc/kernel/ftrace.c
index d5b2739..661210f 100644
--- a/arch/powerpc/kernel/ftrace.c
+++ b/arch/powerpc/kernel/ftrace.c
@@ -74,6 +74,7 @@ ftrace_modify_code(unsigned long ip, unsigned int old, unsigned int new)
  */
 static int test_24bit_addr(unsigned long ip, unsigned long addr)
 {
+ addr = ppc_function_entry((void *)addr);
 
  /* use the create_branch to verify that this offset can be branched */
  return create_branch((unsigned int *)ip, addr, 0);
-- 
1.7.1